fido/src/main/java/com/yubico/yubikit/fido/ctap/FingerprintBioEnrollment.java

```diff
@@ -0,0 +1,496 @@
+     */
+    public static class Context {
+        private final FingerprintBioEnrollment bioEnrollment;
+        @Nullable
+        private final Integer timeout;
+        @Nullable
+        private byte[] templateId;
+        @Nullable
+        private Integer remaining;
+        private final CommandState state = new CommandState();
+
+        public Context(
+                FingerprintBioEnrollment bioEnrollment,
+                @Nullable Integer timeout,
+                @Nullable byte[] templateId,
+                @Nullable Integer remaining) {
+            this.bioEnrollment = bioEnrollment;
+            this.timeout = timeout;
+            this.templateId = templateId;
+            this.remaining = remaining;
+        }
+
+        /**
+         * Capture a fingerprint sample.
+         * <p>
+         * This call will block for up to timeout milliseconds (or indefinitely, if
+         * timeout not specified) waiting for the user to scan their fingerprint to
+         * collect one sample.
+         *
+         * @return None, if more samples are needed, or the template ID if enrollment is
+         * completed.
+         * @throws IOException      A communication error in the transport layer.
+         * @throws CommandException A communication error in the protocol layer.
+         * @throws CaptureError     An error during fingerprint capture.
+         */
+        @Nullable
+        public byte[] capture() throws IOException, CommandException, CaptureError {
+            int sampleStatus;
+            try {
+                if (templateId == null) {
+                    final EnrollBeginStatus status = bioEnrollment.enrollBegin(timeout, state);
+                    templateId = status.getTemplateId();
+                    remaining = status.getRemaining();
+                    sampleStatus = status.getSampleStatus();
+                } else {
+                    final CaptureStatus status = bioEnrollment.enrollCaptureNext(
+                            templateId,
+                            timeout,
+                            state);
+                    remaining = status.getRemaining();
+                    sampleStatus = status.getSampleStatus();
+                }
+
+                if (sampleStatus != FEEDBACK_FP_GOOD) {
+                    throw new CaptureError(sampleStatus);
+                }
+
+                if (remaining == 0) {
+                    return templateId;
+                }
+            } catch (CtapException ctapException) {
+                if (ctapException.getCtapError() == CtapException.ERR_KEEPALIVE_CANCEL) {
+                    bioEnrollment.enrollCancel();
+                    templateId = null;
+                }
+                throw ctapException;
+            }
+
+            return null;
+        }
+
+        /**
+         * Cancels ongoing enrollment.
+         */
+        public void cancel() {
+            state.cancel();
+        }
```

Comment:
The way CommandState and cancel is handled in this class is quite different from how python-fido2 handles it. Should it be different, if so why? If not, which one is correct?

Reply:
I am reviewing this.

Reply:
Updated in [3a98b45](https://github.com/Yubico/yubikit-android/pull/118/commits/3a98b4558ac38974dc388c72e06a6459808e7895). The functionality is now equivalent to the python-fido2 version: the `capture` method has a `CommandState` argument which can be used to cancel the authenticator operation.